File: backend/app/api/mission.py
import asyncio

from fastapi import APIRouter, Depends, HTTPException

from app.api.deps import get_runtime
from app.schemas.mission import MissionSnapshot, MissionUpdate, Stage1HitRequest, Stage1ManualMotionRequest, Stage1PlanRequest, Stage1WrongTargetRequest, Stage2RoundCompleteRequest, Stage3RoundCompleteRequest
from app.schemas.stage2_engagement import Stage2EngagementStatus
from app.schemas.stage3_engagement import Stage3EngagementStatus
from app.services.runtime_state import RuntimeState

router = APIRouter(prefix="/api/mission", tags=["mission"])


@router.get("/status", response_model=MissionSnapshot)
def mission_status(runtime: RuntimeState = Depends(get_runtime)) -> MissionSnapshot:
    return runtime.mission.snapshot()


@router.put("/status", response_model=MissionSnapshot)
async def update_mission(update: MissionUpdate, runtime: RuntimeState = Depends(get_runtime)) -> MissionSnapshot:
    try:
        previous_stage = runtime.mission.state.active_stage
        snapshot = runtime.mission.update(update)
        if snapshot.state.active_stage != previous_stage:
            # A mission-mode transition cannot inherit an autonomous velocity,
            # queued trigger pulse or selected tracker target from its prior mode.
            runtime.command_gateway.stop_motion()
            await runtime.tracking_loop.stop()
            runtime.auto_tracker.stop_tracking()
            if previous_stage == "stage2" or snapshot.state.active_stage == "stage2":
                runtime.stage2_engagement.reset(snapshot.state.stage2_round)
            if previous_stage == "stage3" or snapshot.state.active_stage == "stage3":
                runtime.stage3_engagement.reset(snapshot.state.stage3_round)
            runtime.last_motion_event = ("mission.mode_transition_safed", {"from": previous_stage, "to": snapshot.state.active_stage})
        return snapshot
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc


@router.post("/reset", response_model=MissionSnapshot)
def reset_mission(runtime: RuntimeState = Depends(get_runtime)) -> MissionSnapshot:
    snapshot = runtime.mission.reset()
    runtime.stage2_engagement.reset(snapshot.state.stage2_round)
    runtime.stage3_engagement.reset(snapshot.state.stage3_round)
    return snapshot


@router.put("/stage1/plan", response_model=MissionSnapshot)
def configure_stage1_plan(request: Stage1PlanRequest, runtime: RuntimeState = Depends(get_runtime)) -> MissionSnapshot:
    try:
        return runtime.mission.configure_stage1_plan(request)
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc


@router.post("/stage1/plan/lock", response_model=MissionSnapshot)
def lock_stage1_plan(runtime: RuntimeState = Depends(get_runtime)) -> MissionSnapshot:
    try:
        return runtime.mission.start_stage1()
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc


@router.post("/stage1/start", response_model=MissionSnapshot)
def start_stage1(runtime: RuntimeState = Depends(get_runtime)) -> MissionSnapshot:
    try:
        return runtime.mission.start_stage1()
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc


@router.post("/stage1/hit", response_model=MissionSnapshot)
def record_stage1_hit(request: Stage1HitRequest, runtime: RuntimeState = Depends(get_runtime)) -> MissionSnapshot:
    try:
        return runtime.mission.record_stage1_hit(request)
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc


@router.post("/stage1/wrong-target", response_model=MissionSnapshot)
def record_stage1_wrong_target(request: Stage1WrongTargetRequest, runtime: RuntimeState = Depends(get_runtime)) -> MissionSnapshot:
    try:
        return runtime.mission.record_stage1_wrong_target(request)
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc


@router.post("/manual-motion")
async def manual_motion(request: Stage1ManualMotionRequest, runtime: RuntimeState = Depends(get_runtime)) -> dict:
    result = runtime.command_gateway.send_motion(runtime, request.speed_x, request.speed_y, origin="manual_operator")
    if not result.accepted:
        return {"accepted": False, "reason_codes": result.reason_codes, "detail": result.detail}
    await asyncio.sleep(request.duration_ms / 1000.0)
    stop = runtime.command_gateway.stop_motion()
    return {"accepted": True, "reason_codes": [], "detail": "Manual motion completed through CommandGateway.", "stop_ack": stop.pico_ack}


@router.post("/stage2/round/complete", response_model=MissionSnapshot)
def complete_stage2_round(request: Stage2RoundCompleteRequest, runtime: RuntimeState = Depends(get_runtime)) -> MissionSnapshot:
    # Dry-run/replay scoring remains available for deterministic scoring unit
    # tests.  A live competition cannot supply a free-form hit count.
    if runtime.command_gateway.profile.value != "DRY_RUN":
        raise HTTPException(status_code=409, detail="A2_ENGAGEMENT_EVENT_API_REQUIRED")
    try:
        return runtime.mission.complete_stage2_round(request)
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc


@router.get("/stage2/engagement", response_model=Stage2EngagementStatus)
def stage2_engagement_status(runtime: RuntimeState = Depends(get_runtime)) -> Stage2EngagementStatus:
    return runtime.stage2_engagement.status()


@router.post("/stage2/round/close", response_model=MissionSnapshot)
def close_stage2_round(runtime: RuntimeState = Depends(get_runtime)) -> MissionSnapshot:
    try:
        snapshot = runtime.stage2_engagement.close_round(runtime.mission)
        # Official round boundary: inherit neither velocity, association nor
        # confirmation state into the next three-target pass.
        runtime.command_gateway.stop_motion()
        runtime.auto_tracker.multi_target_tracker.reset()
        runtime.auto_tracker.preferred_target_x = None
        runtime.auto_tracker.preferred_target_y = None
        runtime.association.reset()
        runtime.target_priority.reset()
        runtime.hit_confirmation.reset()
        runtime.last_motion_event = ("mission.stage2_round_safed", {"next_round": snapshot.state.stage2_round})
        return snapshot
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc


@router.post("/stage3/round/complete", response_model=MissionSnapshot)
def complete_stage3_round(request: Stage3RoundCompleteRequest, runtime: RuntimeState = Depends(get_runtime)) -> MissionSnapshot:
    if runtime.command_gateway.profile.value != "DRY_RUN":
        raise HTTPException(status_code=409, detail="A3_ENGAGEMENT_EVENT_API_REQUIRED")
    try:
        return runtime.mission.complete_stage3_round(request)
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc


@router.get("/stage3/engagement", response_model=Stage3EngagementStatus)
def stage3_engagement_status(runtime: RuntimeState = Depends(get_runtime)) -> Stage3EngagementStatus:
    return runtime.stage3_engagement.status()


@router.post("/stage3/round/close", response_model=MissionSnapshot)
def close_stage3_round(runtime: RuntimeState = Depends(get_runtime)) -> MissionSnapshot:
    try:
        snapshot = runtime.stage3_engagement.close_round(runtime.mission)
        runtime.command_gateway.stop_motion()
        runtime.auto_tracker.multi_target_tracker.reset()
        runtime.auto_tracker.preferred_target_x = None
        runtime.auto_tracker.preferred_target_y = None
        runtime.association.reset()
        runtime.target_priority.reset()
        runtime.hit_confirmation.reset()
        runtime.last_motion_event = ("mission.stage3_round_safed", {"next_round": snapshot.state.stage3_round})
        return snapshot
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc

File: frontend/src/api/mission.ts
import type { MissionSnapshot, MissionUpdate, Stage1RangeScore, Stage1Target, Stage2EngagementStatus, Stage3EngagementStatus, Stage3TargetClass } from '../types/mission'

function apiBaseUrl(): string {
  const configured = import.meta.env.VITE_BACKEND_API_URL as string | undefined
  if (configured) return configured.replace(/\/$/, '')
  if (window.location.port && window.location.port !== '5173') return window.location.origin
  return `${window.location.protocol}//${window.location.hostname}:8000`
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${apiBaseUrl()}${path}`, {
    headers: { 'Content-Type': 'application/json', ...init?.headers },
    ...init,
  })
  const body = await response.json()
  if (!response.ok) throw new Error(JSON.stringify(body))
  return body as T
}

export function fetchMissionStatus(): Promise<MissionSnapshot> {
  return request<MissionSnapshot>('/api/mission/status')
}

export function updateMissionStatus(update: MissionUpdate): Promise<MissionSnapshot> {
  return request<MissionSnapshot>('/api/mission/status', {
    method: 'PUT',
    body: JSON.stringify(update),
  })
}

export function resetMissionStatus(): Promise<MissionSnapshot> {
  return request<MissionSnapshot>('/api/mission/reset', { method: 'POST' })
}

export function configureStage1Plan(order: Stage1Target[]): Promise<MissionSnapshot> {
  return request('/api/mission/stage1/plan', { method: 'PUT', body: JSON.stringify({ order }) })
}

export function lockStage1Plan(): Promise<MissionSnapshot> {
  return request('/api/mission/stage1/start', { method: 'POST' })
}

export function recordStage1Hit(target: Stage1Target, scoreAwarded: Stage1RangeScore): Promise<MissionSnapshot> {
  return request('/api/mission/stage1/hit', { method: 'POST', body: JSON.stringify({ target, score_awarded: scoreAwarded }) })
}

export function recordStage1WrongTarget(target: Stage1Target): Promise<MissionSnapshot> {
  return request('/api/mission/stage1/wrong-target', { method: 'POST', body: JSON.stringify({ target }) })
}

export function sendStage1ManualMotion(payload: { speed_x: number; speed_y: number; duration_ms: number }): Promise<{ accepted: boolean; reason_codes: string[]; detail: string }> {
  return request('/api/mission/manual-motion', { method: 'POST', body: JSON.stringify(payload) })
}

export function completeStage2Round(confirmedHits: number): Promise<MissionSnapshot> {
  return request('/api/mission/stage2/round/complete', { method: 'POST', body: JSON.stringify({ confirmed_hits: confirmedHits }) })
}

export function fetchStage2Engagement(): Promise<Stage2EngagementStatus> {
  return request('/api/mission/stage2/engagement')
}

export function closeStage2Round(): Promise<MissionSnapshot> {
  return request('/api/mission/stage2/round/close', { method: 'POST' })
}

export function fetchStage3Engagement(): Promise<Stage3EngagementStatus> {
  return request('/api/mission/stage3/engagement')
}

export function closeStage3Round(): Promise<MissionSnapshot> {
  return request('/api/mission/stage3/round/close', { method: 'POST' })
}

export function completeStage3Round(payload: { enemy_class: Stage3TargetClass; enemy_hit: boolean; friend_hit: boolean }): Promise<MissionSnapshot> {
  return request('/api/mission/stage3/round/complete', { method: 'POST', body: JSON.stringify(payload) })
}
